src/v/archival/segment_reupload.cc

```diff
@@ -0,0 +1,291 @@
+    }
+
+    auto start = _begin_inclusive;
+    model::offset current_segment_end{0};
+    size_t collected_size = 0;
+    while (current_segment_end < _manifest->get_last_offset()) {
+        auto result = find_next_compacted_segment(start);
+        if (result.segment.get() == nullptr) {
+            break;
+        }
+
+        if (unlikely(!_ntp_cfg)) {
+            _ntp_cfg = result.ntp_conf;
+        }
+
+        auto segment_size = result.segment->size_bytes();
+        if (collected_size + segment_size > _max_uploaded_segment_size) {
+            vlog(
+              archival_log.debug,
+              "Compacted segment collect for ntp {} stopping collection, total "
+              "size: {} will overflow max allowed upload size: {}, current "
+              "collected size: {}",
+              _manifest->get_ntp(),
+              collected_size + segment_size,
+              _max_uploaded_segment_size,
+              collected_size);
+            break;
+        }
+
+        // For the first compacted segment found, begin offset needs to be
+        // re-aligned if it falls inside manifest segment.
+        if (_segments.empty()) {
+            _begin_inclusive = result.segment->offsets().base_offset;
+            align_begin_offset_to_manifest();
+        }
+        _segments.push_back(result.segment);
+        current_segment_end = result.segment->offsets().committed_offset;
+        start = current_segment_end + model::offset{1};
+        collected_size += segment_size;
+    }
+
+    if (current_segment_end >= replace_boundary) {
+        _can_replace_manifest_segment = true;
+    }
+
+    if (current_segment_end >= _manifest->get_last_offset()) {
+        vlog(
+          archival_log.debug,
+          "Compacted segment collect for ntp {} offset {} advanced "
+          "ahead of manifest, clamping to {}",
+          _manifest->get_ntp(),
+          current_segment_end,
+          _manifest->get_last_offset());
+        _end_inclusive = _manifest->get_last_offset();
+    } else {
+        // Align the end offset to the nearest segment ending in manifest.
+        it = _manifest->segment_containing(current_segment_end);
+        if (it == _manifest->end()) {
+            // current_segment_end is in a gap in the manifest.
+            if (current_segment_end >= _manifest->get_start_offset().value()) {
+                vlog(
+                  archival_log.info,
+                  "Compacted segment collect for ntp {}: collection ended at "
+                  "gap in manifest: {}",
+                  _manifest->get_ntp(),
+                  current_segment_end);
+
+                // try to fill the manifest gap with the data locally available.
+                _end_inclusive = current_segment_end;
+            }
+            return;
+        }
+
+        // If the compacted segment end is not aligned to manifest segment, then
+        // pull back to the end of the previous segment.
```

Comment:
Can we read up to some offset in the misaligned segment in this case?

Reply:
From what I can see the way to avoid overlap with manifest segment at the end of collection is to move the end of collected segments back to nearest manifest segment boundary.

eg manifest has two segments: 20-29,30-39 and compacted segment collection starts at 20 and ends at offset 35.

If we upload 20-35 then we will have overlap from 30-35, and we cannot progress collection ahead of 35, so the only option is to set the end of collection back to 29, and upload data 20-29 replacing the first manifest segment.

If there is another option please let me know.

Reply:
got it, makes sense 
so here you're not doing conversion from redpanda offset to file offset since it's done inside archival_policy. correct?

Reply:
yes right, file offset conversion will be done as part of #6455 draft PR here: https://github.com/redpanda-data/redpanda/blob/2636fd4565e22481301dcecdd6a6180972a32550/src/v/archival/segment_reupload.cc#L258